syft/frameworks/torch/tensors/interpreters/chinese_remainder_theorem.py

```diff
@@ -0,0 +1,174 @@
+    """
+
+    def __init__(self, residues: dict = None, owner=None, id=None, tags=None, description=None):
+        super().__init__(owner=owner, id=id, tags=tags, description=description)
+
+        # check that all the modulos are pairwise coprime
+        if residues is not None:
+            for pair in itertools.combinations(residues.keys(), r=2):
+                assert (
+                    math.gcd(pair[0], pair[1]) == 1
+                ), f"{pair[0]} and {pair[1]} are not coprime, you cannot build a CRTTensor with these as modulos"
+
+        if residues is not None:
+            res_shape = next(iter(residues.values())).shape
+            self.child = {}
+            for f, r in residues.items():
+                assert isinstance(r.child, FixedPrecisionTensor)
+                assert f == r.child.field
+                self.child[f] = r
+                assert (
+                    r.shape == res_shape
+                ), "All residue tensors of CRTTensor must have the same shape"
+
+    def __str__(self):
+        type_name = type(self).__name__
+        out = f"[" f"{type_name}]"
+        for mod, res in self.child.items():
+            out += f"\n\t -> {mod}: {res}"
+        return out
+
+    def __repr__(self):
+        return self.__str__()
+
+    @property
+    def shape(self) -> torch.Size:
+        return next(iter(self.child.values())).shape
+
+    @overloaded.method
+    def eq(self, self_, other_):
+        assert (
+            self_.keys() == other_.keys()
+        ), "Cannot compare 2 CRT that don't have the same modulos"
+
+        result = torch.ones(self.shape).fix_prec(precision_fractional=0)
+        for mod in self_.keys():
+            result *= (self_[mod] == other_[mod]).long()
+        
+        dict_result = {}
+        for mod in self_.keys():
+            dict_result[mod] = result
+
+        return dict_result
+
+    __eq__ = eq
+
+    @overloaded.method
+    def add(self, self_, other):
+        assert self_.keys() == other.keys(), "Cannot add 2 CRT that don't have the same modulos"
+
+        result = {}
+
+        for mod in self_.keys():
+            result[mod] = self_[mod] + other[mod]
+
+        return result
+
+    __add__ = add
+
+    @overloaded.method
+    def sub(self, self_, other):
+        assert (
+            self_.keys() == other.keys()
+        ), "Cannot subtract 2 CRT that don't have the same modulos"
+
+        result = {}
+
+        for mod in self_.keys():
+            result[mod] = self_[mod] - other[mod]
+
+        return result
+
+    __sub__ = sub
+
+    @overloaded.method
+    def mul(self, self_, other):
+        assert (
+            self_.keys() == other.keys()
+        ), "Cannot multiply 2 CRT that don't have the same modulos"
+
+        result = {}
+
+        for mod in self_.keys():
+            result[mod] = self_[mod] * other[mod]
+
+        return result
+
+    __mul__ = mul
+
+    def solve_system(self):
```

Comment:
not sure where this is being called, so may already be the case, but most of this can be precomputed and reused. reconstruction in that case then boils down to a weighted sum.

Reply:
I see what you mean. Should I have the coefficients of this reconstruction as attributes computed when the tensor is instantiated and used when needed?

Reply:
do you have an example of where/how the CRT tensors are created? the ideal point to compute and cache these is where the moduli are provided (eg along these lines: https://github.com/tf-encrypted/tf-encrypted/blob/master/tf_encrypted/tensor/int100.py#L757-L760)

Reply:
btw, i've looked quite a lot at CRT tensors and happy to discuss!

Reply:
We still don't use CRT tensors so I don't have any real life example to show you... In test_chinese_remainder.py (in this PR), you can see how I create these tensors for the moment (I create the tensors of residues and build a CRTTensor instance with them).
Actually, it's still a bit blurry to me how and when this kind of representation is used, how it is created in the first place, etc. So I might come to ask you some questions if you don't mind :)